Horizons-Games/Axolotl-Engine
# feat: loading scripts logic

This merge request adds a `SceneLoadingScript` and a demo loading screen courtesy of @ericdacal.

# How does it work?

`SceneLoadingScript` exposes a single member, which is the real scene we want to load. It will automatically format the path, so setting the string field to "Example" will result in an attempt to load `Lib/Scenes/Example.axolotl`. It has a constant, hard-coded path to what is assumed to be the default loading scene, `LoadingScreen`. This can be changed later if needed (say, by adding more loading screens and showing one randomly). On starting, the script verifies that both the constant loading screen and the dynamic scene to be loaded exist.

The script needs to be called by another one; in this MR I've updated the Main Menu scene so the "Play" button calls this new script. When the method `SceneLoadingScript::StartLoad` is called, the script loads the loading scene, waits 6 frames so that `ModuleRender` can tick (render is blocked during loading) and starts the load of the actual scene, using the callbacks the scene loading method provides.

# How can you test?

You'll need a sort-of functional Game flow to test this, and remember that only Main Menu has been updated.

1. Load `00_MainMenu_VS3.axolotl`, `LoadingScreen.axolotl` and `__LEVEL1_ALFA.axolotl`, in whatever order you want. This will populate the Lib folder.
2. Once the three scenes have been loaded, choose "Build" on the main menu of the engine, select `00_MainMenu_VS3.axolotl` as the starting scene, choose to regenerate the .zip file and start the build (compilation configuration doesn't matter, up to you).
3. Grab a coffee, call your family, do some exercise, whatever you feel like.
4. Once the build is done, start the game and press "Play". You should see a screenshot of the sewers (the temporary loading screen) and after a while the level one should load.

# Engine changes

Engine-side required mostly guards to prevent updates from happening while the scene was loaded. ~Coolest addition, even though it's probably useless outside this use case, is that an optional parameter has been added to the `Scheduler::ScheduleTask` method to set a delay (in frames) for the task to run. This was done, as stated before, so the `ModuleRender` could run for a bit.~ Cherrypicked in #766 instead.

# What to do?

The scenes would need to be updated so all transitions use this logic, eithe by me or someone else in this MR or another.

We also need to see how much stuff we can put on the loading screen. The quickest way to make this whole thing work was to prevent a bunch of modules from updating during load, but we can try to roll those changes back and allow some functionality so we can have some visual feedback.
